Player Visual Controller

SwiffterAnim opened the pull request:
Added the visual controller that swaps sprites when player gets damaged. Temp sprites for now.
There will be 5 sprites in total:
Uncracked
25 % Cracked
50% Cracked
75% Cracked and
100% Cracked.

SwiffterAnim committed:
Player Visual Controller

Added the visual controller that swaps sprites when player gets damaged.
Temp sprites for now.
There will be 5 sprites in total:
Uncracked
25 % Cracked
50% Cracked
75% Cracked and
100% Cracked.

SwiffterAnim committed:
Deactivating Hooks

I added the logic of the deactivation of the hooks after the player jumps on them. I wasn't sure if I added it to the correct place but I just wanted to advance with it.

SwiffterAnim committed:
Fixes Suggested in our Previous Meeting

I made all the changes with your teachings and suggestions from our last meeting. Created CollisionController, LifeController, HookController, and changed everything accordingly.

SwiffterAnim committed:
Tried implemeting Die() and failed haha

I created a method Die() but unfortunately it's not working.
The way I was trying to make it work is by disabling the component PlayerMovementController, but even though it's working (it is disabling) somehow the controllers still work, the script is still having an effect on the game. Destroying the component worked, but I was trying to avoid that.
Need to leave this for now unfortunately. Cheers!

omidizadi commented on Assets/Modules/Movement/Scripts/PlayerMovementController.cs:
```diff
@@ -66,6 +77,7 @@ private void OnAim(InputValue value)
     private void OnJump(InputValue value)
     {
         rb.isKinematic = false;
```
The reason that this script works even when it's disabled, is because of the way the "PlayerInput"  has been set up.
![image](https://github.com/SwiffterAnim/Bounce-Escape/assets/47353647/1eef9078-f247-4440-b480-fe66ce8b4188)

As you can see, the behaviour is set to "Send Message". This means that the player input component will look for any method that is named like "OnJump" "OnAim" etc. and exists in one of the components attached to the GameObject. With this behaviour, even if the script is disabled, the method will be called by PlayerInput.

omidizadi replied:
Instead of disabling the PlayerMovementController, you can introduce a new method inside PlayerMovementController called `DisableInput` and then there you should do the following:
First create a field so that you can reference the player input:
![image](https://github.com/SwiffterAnim/Bounce-Escape/assets/47353647/39070964-337e-4438-bb14-86d4f1d3f819)
Then create the disable input method:
![image](https://github.com/SwiffterAnim/Bounce-Escape/assets/47353647/d55ae68f-07df-46a5-84bb-916f054a7806)
Then instead of disabling the script, call this method:
![image](https://github.com/SwiffterAnim/Bounce-Escape/assets/47353647/21cb7a5f-d62c-4c30-9419-332cac35a3f0)

SwiffterAnim replied:
Ok this kinda worked, so that made that I can't control the character anymore, but the character is still attracted by the hooks... So I tried, again, to disable the whole playermovementcontroller script, in order for it NOT to be attracted by the hooks, and it still gets attracted somehow, the disabling is working, I see it in the Inspector, but it still gets "hooked". The script that makes it hook is in the PlayerMovementController, the method OnTriggerEnter2D, and it checks if the hook has "hooknessActivated" it hooks the character to the hook.

I still don't understand how it still gets attracted if I'm disabling the whole script haha

omidizadi replied:
hmm weird, I just tested it with a joystick and it seems that after dying the player doesn't get hooked.

https://github.com/user-attachments/assets/bfe61820-4514-4d55-bb71-afdab5ef0b8b

omidizadi replied:
I should mention that my Unity version is 2022.3.9f1. I highly doubt that it matters but maybe, just "maybe" that could be a bug in your Unity version.

SwiffterAnim replied:
Ah yes on the latest version of the build that's already fixed! I fixed it with a boolean "isAlive" to it, and made it work without disabling the script. That was part of my Question 1 on the mentor chat for reference =)x

omidizadi commented on Assets/Modules/Obstacles/Scripts/ObstacleHookController.cs:
```diff
@@ -0,0 +1,56 @@
+        sr = GetComponent<SpriteRenderer>();
+    }
+    
+    void Start()
```
If you don't need a Unity method (Start, Awake, Update, LateUpdate, etc) always remove them. They add unnecessary overhead when the project gets bigger and bigger.

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/Obstacles/Scripts/ObstacleHookController.cs:
```diff
@@ -0,0 +1,56 @@
+        sr.color = tmp;
+        obstacleEntity.hooknessActivated = true;
+        gameObject.AddComponent<CircleCollider2D>();
+        CircleCollider2D myCollider = gameObject.GetComponent<CircleCollider2D>();
```
You are calling `GetComponent<CircleCollider2D>()` in a coupld of places in this script. Instead, just create a [SerializeField] and reference the collider.

SwiffterAnim replied:
I think this one I can't SerializeField because it's destroyed at some point, am I wrong?
So as far as I understand, serialized fields, I can assign them in the Inspector, but if I destroy that component, when I recreate it in the future (which is what I'm doing when I "ReactivateHook()", it won't be assigned in the serialized field anymore right?

omidizadi replied:
Yes exactly! They refer to an instance of that class that you define and if that instance gets destroyed, you need to re-assign it. For now, let's keep it like this. We can think about how to do it properly in future :)

omidizadi commented on Assets/Modules/Player Visuals/Scripts/PlayerVisualController.cs:
```diff
@@ -0,0 +1,28 @@
+
+    public void CrackCristalBall(int crackIndex)
+    {
+        sr.sprite = crackStages[crackIndex];        
```
Good job 👍 This script is following single responsibility principle exactly as it should 😄

omidizadi commented on Assets/Modules/Player Visuals/Scripts/RingCollider2D.cs:
```diff
@@ -0,0 +1,29 @@
+
+public class RingCollider2D : MonoBehaviour
+{
+    [SerializeField] int numSegments = 16;
```
Do not forget `private` modifier for variables and methods.

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/PlayerLifeManager.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        playerCollisionController.OnCollisionEnter2DEvent += OnCollisionEnter2DEvent;
+
+        if(!gameObject.GetComponent<PlayerMovementController>().enabled)
```
Same comment about `GetComponent<PlayerMovementController>()` you can just create a serializefield and cache it

SwiffterAnim replied:
Done

omidizadi commented on Assets/Modules/PlayerLifeManager.cs:
```diff
@@ -0,0 +1,98 @@
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class PlayerLifeManager : MonoBehaviour
```
It's better to be called `PlayerLifeController` since the player object only needs this.

SwiffterAnim replied:
Done.

SwiffterAnim committed:
Made Die work.

Fixed your previous notes and made it work with the isAlive bool. Going to tackle other tasks!

SwiffterAnim committed:
Time Manager

Some minor fixes in some notes of yours that I missed and did the Time Manager.

SwiffterAnim committed:
PlayerAimVisualController

Added the aim.
Not sure if I prefer this or what I had before, which was just a straight line. Was a bumpy road, I'm not sure about all of what's there. Especially that variable "divisionForLine"...
Otherwise, it's kinda working haha

SwiffterAnim committed:
Enemy Things

All Enemy related tasks.
EnemyMoveController
EnemySpawnManager
ProjectileSpawnController
ProjectileEntity.
Please refer to finishing daily notes (day 2) for caviats and explanations and questions!
Thank you!

omidizadi reviewed (approved):
I think this branch is in a good shape :) you can merge it and create a new PR for the upcoming changes. Well done! 👍

SwiffterAnim committed:
Organisation/Bugs/Fixes

This are a bunch of additions with your notes from this week's development.
Started testing the visual for the player.

SwiffterAnim committed:
(FAILED) Attempt to liquid sim with shaders

Wanted to commit in case you wanted to check, but I failed hehe will move on.

SwiffterAnim committed:
(FAILED) ProjectileVelocity to Player

And other bits and bobs.

SwiffterAnim committed:
Added a bit of fun breaking

When the character dies, now it explodes into many pieces =D
